### ISSUE TITLE
Improving messages expected in defining data types using lowcase identifier and correspondencing functions, errors in order to avoid ambiguity.

### ISSUE THREAD

**Issue by ghost** [idris-lang/Idris2#877]
# Steps to Reproduce

Version 0.2.1-b0e6cdc16

```idris
module Bug
data TeXNat = zero | succ TeXNat
dts : TeXNat -> String
dts n = case n of
  zero => "zero"
```

# Expected Behavior

Report can not define data types with lowercase identifier.

# Observed Behavior

```
alissa@stackage ~/scratch % idris Bug.idr
     ____    __     _         ___                                           
    /  _/___/ /____(_)____   |__ \                                          
    / // __  / ___/ / ___/   __/ /     Version 0.2.1-b0e6cdc16
  _/ // /_/ / /  / (__  )   / __/      https://www.idris-lang.org           
 /___/\__,_/_/  /_/____/   /____/      Type :? for help                     

Welcome to Idris 2.  Enjoy yourself!
Bug> :l Bug.idr
Loaded file Bug.idr
Bug> 
Bye for now!
```

**Comment by buzden** [comment]
What's wrong with lowercase data constructors? You can even match on them:

```idris
module LowerCaseData

%default total

data TeXNat = zero | succ TeXNat
dts : TeXNat -> String
dts n = case n of
  LowerCaseData.zero   => "zero"
  LowerCaseData.succ _ => "non-zero"
```

I think everything works as expected.

**Comment by alissa42** [comment]
> What's wrong with lowercase data constructors? You can even match on them:
> 
> ```idris
> module LowerCaseData
> 
> %default total
> 
> data TeXNat = zero | succ TeXNat
> dts : TeXNat -> String
> dts n = case n of
>   LowerCaseData.zero   => "zero"
>   LowerCaseData.succ _ => "non-zero"
> ```
> 
> I think everything works as expected.

That may become the problem of show the messages for ambiguous occurrences of `zero`, for syntax to use lowercase identifier to distinct type variables from data constructors?

**Comment by buzden** [comment]
> That may become the problem of show the messages for ambiguous occurrences of `zero`, for syntax to use lowercase identifier to distinct type variables from data constructors?

Yes, it may. But I think that this is not a problem of what identifiers are allowed in data declarations. So, maybe error messages should be improved, for example printing the point where a name is defined when using the name in an error message.

**Comment by gallais** [comment]
We should at least warn users when they define a datatype with
lowercase constructors as that breaks a lot of the assumptions
made in the compiler.

**Comment by gallais** [comment]
Fixed by #1207
